client/LaunchVM.py

```diff
@@ -0,0 +1,181 @@
+    except grpc.RpcError as e:
+        print("Failed to get Launch Bundle: {}".format(e), file=sys.stderr)
+        channel.close()
+        exit(1)
+
+    # add godh and session file to XML
+    ls = soup.find("launchSecurity")
+    cert_tag = soup.new_tag("dhCert")
+    cert_tag.string = response.GuestOwnerPublicKey
+    ls.append(cert_tag)
+
+    session_tag = soup.new_tag("session")
+    session_tag.string = response.LaunchBlob
+    ls.append(session_tag)
+
+    xml = str(soup.domain)
+
+    # define the domain from XML
+    dom = hv.defineXML(xml) 
+    dom.createWithFlags(libvirt.VIR_DOMAIN_START_PAUSED)
+
+    # get launch measurement
+    sevinfo = dom.launchSecurityInfo()
+
+    expected_digest = get_expected_digest(ovmf_path, initrd_path, kernel_path, cmdline)
+
+    request = SecretRequest(LaunchMeasurement = sevinfo['sev-measurement'], \
+            LaunchId = response.LaunchId, \
+            Policy = sevinfo['sev-policy'], \
+            ApiMajor = sevinfo['sev-api-major'], \
+            ApiMinor = sevinfo['sev-api-minor'], \
+            BuildId = sevinfo['sev-build-id'], \
+            FwDigest = expected_digest, \
+            LaunchDescription = "test launch",
+            SecretRequests = [\
+                    RequestDetails(Guid = "0a46e24d-478c-4eb1-8696-113eeec3aa99", \
+                        Format = "json", \
+                        SecretType = "bundle", \
+                        Id = "keyset1" )] \
+            )
+   
+    try:
+        response = client.GetSecret(request)
+    except grpc.RpcError as e:
+        print("Failed to get Launch Secret: {}".format(e), file=sys.stderr)
+        channel.close()
+        exit(1)
+
+
+    params = {"sev-secret": response.LaunchSecretData,
+              "sev-secret-header": response.LaunchSecretHeader}
+    dom.setLaunchSecurityState(params, 0)
+    dom.resume()
+
+    hv.close()
+
+# calculate the expected launch digest 
+# this will be verified via he measurement
+# so it's fine for the CSP to calculate 
+# this here
+def get_expected_digest(ovmf, initrd, kernel, cmdline):
+
+    fh = open(initrd, 'rb')
+    h = hashlib.sha256(fh.read())
+    initrd_hash = h.digest()
+
+    fh = open(kernel, 'rb')
+    h = hashlib.sha256(fh.read())
+    kernel_hash = h.digest()
+
+    cmdline = bytearray(cmdline, encoding='utf8')
+
+    # remove trailing `\n` if present
+    cl = len(cmdline)
+    if (cmdline[cl-1] == 0xa):
+        cl -= 1
+
+    # add trailing '\0'
+    cmdline[cl:cl+1] = (0).to_bytes(1, byteorder='little')
+    h = hashlib.sha256(cmdline)
+    cmdline_hash = h.digest()
+
+    h = hashlib.sha256();
+    fh = open(ovmf, 'rb');
+    h.update(fh.read());
+    hashes = bytearray()
+    hashes[0:16]= UUID('{9438d606-4f22-4cc9-b479-a793d411fd21}').bytes_le
+    l = 16 + 2                   # GUID + 2 byte length
+    hashes[16:18] = l.to_bytes(2, byteorder='little')
+
+    hashes[l:l+16] = UUID('{97d02dd8-bd20-4c94-aa78-e7714d36ab2a}').bytes_le
```

Comment:
I find this way of writing things a bit error prone. I would suggest adding a helper function (or even helper class) that takes a string or a number and adds it to the hash.

Reply:
Yes, you're right. In a separate future PR I think we should change to Python's [ctypes.Structure](https://docs.python.org/3/library/ctypes.html?highlight=structure#ctypes.Structure) (with `_pack_=1`) to describe such a struct ("copying" the struct definition from QEMU) and fill it in.